File: 01. DataStructures/02. Stack/Basic Operation/3. Stack using Queue/Stack_Queue.java
import java.util.*;

public class Stack_Queue { 
	public static void main(String[] args) {
		Scanner sc = new Scanner(System.in);
		Stack_Q st = new Stack_Q();

		boolean flag = true;
		int val = 0;
		while (flag) {

			System.out.println("1. Push item in Stack");
			System.out.println("2. Pop Item out of Stack");
			System.out.println("3. Current Size of Stack");
			System.out.println("4. Peek()");
			System.out.println("5. View Stack");
			System.out.println("6. Exit");

			System.out.println("Enter Choice");
			int choice = sc.nextInt();
			switch (choice) {

			case 1:
				System.out.println("Enter value");
				val = sc.nextInt();
				st.push(val);
				break;

			case 2:
				System.out.println(st.pop());
				break;

			case 3:
				System.out.println(st.count());
				break;

			case 4:
				System.out.println(st.peek());
				break;

			case 5:
				st.view();
				break;

			case 6:
				flag = false;
				break;

			default:
				System.out.println("invalid choice");
			}// switch
			System.out.println();
		} // while

	}

}

class Stack_Q {
	Queue<Integer> q1;// primary queue as stack
	Queue<Integer> q2;// secondary queue for push
	int count;

	public Stack_Q() {
		q1 = new LinkedList<>();
		q2 = new LinkedList<>();
		count = 0;
	}

	public boolean isEmpty() {
		return q1.isEmpty();
	}

	public int count() {
		return count;
	}

	public int peek() {
		return q1.peek();
	}

	public void push(int val) {
		q2.add(val);
		while (!q1.isEmpty()) {
			q2.add(q1.remove());
		}
		Queue<Integer> temp = q1;
		q1 = q2;
		q2 = temp;
		count++;
	}

	public int pop() {
		try {
			count--;
			return q1.remove();
		} catch (Exception e) {
			System.out.println("Exception occurs " + e.getMessage());
			return -999;
		}
	}

	public void view() {
		System.out.println(q1);
	}

}

File: 01. DataStructures/02. Stack/Basic Operation/3. Stack using Queue/Stack_Queue.js
// - Enqueue()  Add item to the queue from the REAR.
// - Dequeue()  Remove item from the queue from the FRONT.
// - isFull()  Check if queue is full or not.
// - isEmpty()  Check if queue empty or not.
// - count() Get number of items in the queue.
// - peek() return front elemenet in the queue(line).
// 1 2 3 4 5 6 7 8 9 10


class Stack {
    constructor(size){
        this.stack = [];
    }
    
    isEmpty(){
        return this.stack.length === 0 ? true : false;
    }
    push(data) {
        return this.stack.push(data);
    }
    pop(){
        return this.stack.pop();
    }
    size() {
        return this.stack.length;
    }
}

class Queue {
    constructor(size) {
        this.stackA = new Stack();
        this.stackB = new Stack();
        this.size = size;
        // [2,3,4,5,6]
        // [] []
    }

    enqueu(data) {
        // console.log(this.stackA.size(), this.stackB.size(), this.size);
        if(this.stackA.size() + this.stackB.size() < this.size) {
            this.stackA.push(data);
        }
    }

    dequeue() {
        if(this.stackB.isEmpty()) {
            while(!this.stackA.isEmpty()) {
                this.stackB.push(this.stackA.pop());
            }
        }

        this.stackB.pop();
    }

    peek() {
        if(this.stackB.isEmpty()) {
            return this.stackA.stack[0];
        } else {
            return this.stackB.stack[this.stackB.stack.length - 1];
        }
    }

    isEmpty() {
        if(this.stackA.isEmpty() && this.stackB.isEmpty()) {
            return true;
        } else {
            return false;
        }
    }

    isFull() {
        if(this.stackA.isEmpty() && this.stackB.isEmpty()) {
            return false;
        } else {
            return true;
        }
    }

    count() {
        return this.stackA.size() + this.stackB.size();
    }
}

// 3,10,101,52,4
const myQueue = new Queue(5);
myQueue.enqueu(1); 
myQueue.enqueu(2);
myQueue.enqueu(3);
myQueue.enqueu(4);
myQueue.dequeue();
myQueue.dequeue();
myQueue.dequeue();
myQueue.dequeue();
myQueue.enqueu(5);
myQueue.enqueu(5);
myQueue.enqueu(5);
myQueue.enqueu(5);
myQueue.enqueu(5);
myQueue.dequeue(5);
myQueue.enqueu(5);

console.log(myQueue.count());

